probe/ebpf/flow.c

```diff
@@ -99,11 +100,19 @@ static inline void fill_transport(struct __sk_buff *skb, __u8 protocol, int offs
 
 	switch (protocol) {
 		case IPPROTO_UDP:
-			//len -= sizeof(struct udphdr);
+			offset+=sizeof(struct udphdr);
+			len -= sizeof(struct udphdr);
 			fill_payload(skb, offset, flow, len);
 			break;
 		case IPPROTO_TCP:
-			//len -= sizeof(struct tcphdr);
+			layer->ab_syn=((load_word(skb,offset+2*sizeof(__be16)+2*sizeof(__be32))>>6)&0x00000001)==1?tm:0;
```

Comment:
may load the 16bits word once and test it with static flags
flags := load_word(skb,offset+12)
layer->ab_syn := (flags & 0x02) > 0 ? tm : 0;
...

Reply:
will load it once, though I did not understand why offset+12? can u elaborate a little further plz?

Reply:
look at tcp header, offset point to the start of the header, you need a 12 bytes displacement to access tcp flags fields.    Or why offset+2*sizeof(__be16)+2*sizeof(__be32) ?

Reply:
oh ok it's the same;) sorry didn't calculate it right before

Reply:
Another question if you don't mind:
assuming we are working with Little-Endians, according to http://ttcplinux.sourceforge.net/documents/one/structure/tcphdr.html in order to get the value of the syn flag we should do:
flags=load_byte(skb,offset+14) -- 14 because little endian flips the order between the flags and the window size
layer->ab_syn := (flags & 0x40) > 0 ? tm : 0; -- syn is the seventh bit
Am I wrong here? if we load the 16bit word as you suggested we will still have to roll it by 8 bits to get to the same point...

Reply:
This should work: 

```
layer->ab_syn = (load_byte(skb, offset + 14) & 0x2) > 0 ? tm : 0;
```